---
title: "Single Cell RNAseq Data Processing & Analysing Course"
author: "Sumukh Deshpande"
#full-width: true
---

---
## Day 4 - Cell-type Identification and Trajectory Analysis in Seurat

---

### Recap from Day 3
---
---

- On 27th Feb, we covered:
  - Integrating single-cell datasets.
  - Differential Gene Expression analysis and marker selection.
  - Cell-type annotation using CellMarker.

- Today we will cover:
  - Cell-type annotation using SingleR.
  - Trajectory analysis (pseudotime analysis) using monocle3

---
---
### Cell-type annotation using SingleR
---
---

**What is SingleR?**

- SingleR is an automatic annotation method for single-cell RNA sequencing (scRNAseq) data.
- Given a reference dataset of samples (single-cell or bulk) with known labels, it labels new cells from a test dataset based on similarity to the reference.
- The burden of manually interpreting clusters and defining marker genes only has to be done once.
- This biological knowledge can be propagated to new datasets in an automated manner.

---
#### Using built-in references
---

- The easiest way to use SingleR is to annotate cells against built-in references.
- The `celldex` R package provides access to several reference datasets.

**General-purpose references**
- Human primary cell atlas (HPCA): Publicly available microarray datasets derived from human primary cells.
- Blueprint/ENCODE: Bulk RNA-seq data for pure stroma and immune cells generated by Blueprint and ENCODE projects.
- Mouse RNA-seq: Collection of mouse bulk RNA-seq data sets downloaded from the gene expression omnibus.

**Immune references**
- Immunological Genome Project (ImmGen): Consists of microarray profiles of pure mouse immune cells.
- Database of Immune Cell Expression/eQTLs/Epigenomics (DICE): The `DICE` reference consists of bulk RNA-seq samples of sorted cell populations.
- Novershtern hematopoietic data: The Novershtern reference (previously known as Differentiation Map) consists of microarray datasets for sorted hematopoietic cell populations from GSE24759.
- Monaco immune data: The Monaco reference consists of bulk RNA-seq samples of sorted immune cell populations from GSE107011.

---
#### Downloading reference dataset
---

- Since the dataset we are working with is the NSCLC scRNAseq dataset, we will be working with HPCA dataset.
- To download the reference data, use the `fetchReference()` function.

```
# Load the celldex package

library(celldex)

# Download the reference database

ref <- fetchReference("hpca", "2024-02-26")
```

SingleR requires the seurat object to be converted to `SingleCellExperiment` object. Therefore, we will be doing the conversion step at this point.

```
# Convert Seurat object to SingleCellExperiment object

allsamplesUMAP.SCE <- as.SingleCellExperiment(allsamplesUMAP)
```

Run the SingleR function to perform mapping of target dataset onto the reference dataset.

```
pred.allsamples <- SingleR(test = allsamplesUMAP.SCE, ref = ref, assay.type.test=1,
                           labels = ref$label.main)
```

**What does `SingleR` function do ?**

- It assigns cell types to individual cells by comparing their gene expression profiles to reference datasets with known cell type annotations.
- How `SingleR` Works:
  1. **Input Data**: Takes a query dataset containing gene expression data for single cells.
  2. **Reference Dataset**: Uses a reference dataset with pre-annotated cell types from celldex, such as HumanPrimaryCellAtlasData.
  3. **Correlation Analysis**: Compares the gene expression of each cell in the query dataset to the reference profiles using Spearman correlation or other similarity measures.
  4. **Scoring & Classification**: Assigns the most likely cell type based on the highest similarity score.
  5. **Fine-tuning**: Iteratively refines the annotations by removing ambiguous labels and reassigning cells.

Type in `pred.allsamples` to see the output of the analysis:

```
print(pred.allsamples)
```

Now we will need to convert the row names to cell_ID column and will create a new dataframe with a name `pred.allsamples.df` and remove the row names from the dataframe.

```
pred.allsamples.df <- data.frame(cell_ID = row.names(pred.allsamples), pred.allsamples)
rownames(pred.allsamples.df) <- NULL
```

Next, select the `cell_ID` and `labels` columns and store the columns in a new dataframe.

```
pred.allsamples.df_labels <- pred.allsamples1.df[,c(1,ncol(pred.allsamples1.df)-2)]
```

Create a copy of `allsamplesUMAP` object and call it `allsamplesUMAP1` so that any changes that are made does not affect the original object.

```
allsamplesUMAP1 <- allsamplesUMAP
```

Now, add `cell_ID` column to the `allsamplesUMAP1` seurat object which is copied from the row names in the metadata.

```
allsamplesUMAP1@meta.data <- data.frame(cell_ID = row.names(allsamplesUMAP1@meta.data), allsamplesUMAP1@meta.data)
```

Also, create another column called `predicted_cell_type_SingleR` column in the metadata of `allsamplesUMAP1` object. Dont worry. The values in this column will get replaced with the predicted cell types using SingleR.
